sdk/compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde.hs

```diff
@@ -439,6 +429,61 @@ removeOpenFile miState home file = do
   unsafeIOToSTM $ logInfo miState $ "Removed open file " <> unDamlFile file <> " from " <> unPackageHome home
   onOpenFiles miState home $ Set.delete file
 
+-- Logic for moving open files between subIDEs when packages are created/destroyed
+
+-- When a daml.yaml is removed, its openFiles need to be distributed to another SubIDE, else we'll forget that the client opened them
+-- We handle this by finding a new home (using getSourceFileHome) for each open file of the current subIDE, and assigning to that
+-- We also send the open file notification to the new subIDE(s) if they're already running
+handleRemovedPackageOpenFiles :: MultiIdeState -> PackageHome -> IO ()
+handleRemovedPackageOpenFiles miState home = withIDEsAtomic_ miState $ \ides -> do
+  let ideData = lookupSubIde home ides
+      moveOpenFile :: SubIDEs -> DamlFile -> STM SubIDEs
+      moveOpenFile ides openFile = do
+        -- getSourceFileHome caches on a directory level, so won't do that many filesystem operations
+        newHome <- getSourceFileHome miState $ unDamlFile openFile
+        let newHomeIdeData = lookupSubIde newHome ides
+            newHomeIdeData' = newHomeIdeData {ideDataOpenFiles = Set.insert openFile $ ideDataOpenFiles newHomeIdeData}
+        -- If we're moving the file to a disabled IDE, it should get the new warning
+        when (ideDataDisabled newHomeIdeData') $ traverse_ (sendClientSTM miState) $ disableIdeDiagnosticMessages newHomeIdeData'
+        forM_ (ideDataMain newHomeIdeData) $ \ide -> do
```

Comment:
Do we want to send these messages even if the ide is disabled?

Reply:
Yes, disabled essentially means disallowing the IDE to start. Usually, the IDE is killed before being disabled, so `ideDataMain` will be `Nothing` when `ideDataDisabled` is true, but we handle them separately here to be correct